jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java

```diff
@@ -104,15 +128,285 @@ public Object getAttribute(String name)
+            return _attributes.getAttributeNames();
         }
 
         @Override
         public void clearAttributes()
         {
             _attributes.clearAttributes();
         }
+
+        @Override
+        public int hashCode()
+        {
+            return _attributes.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return _attributes.equals(obj);
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by a {@link ConcurrentHashMap}.
+     */
+    class Mapped implements Attributes
+    {
+        private final java.util.concurrent.ConcurrentMap<String, Object> _map = new ConcurrentHashMap<>();
+        private final Set<String> _names = Collections.unmodifiableSet(_map.keySet());
+
+        public Mapped()
+        {
+        }
+
+        public Mapped(Mapped attributes)
+        {
+            _map.putAll(attributes._map);
+        }
+
+        @Override
+        public Object removeAttribute(String name)
+        {
+            return _map.remove(name);
+        }
+
+        @Override
+        public Object setAttribute(String name, Object attribute)
+        {
+            if (attribute == null)
+                return _map.remove(name);
+
+            return _map.put(name, attribute);
+        }
+
+        @Override
+        public Object getAttribute(String name)
+        {
+            return _map.get(name);
+        }
+
+        @Override
+        public Set<String> getAttributeNames()
+        {
+            return _names;
+        }
+
+        @Override
+        public void clearAttributes()
+        {
+            _map.clear();
+        }
+
+        public int size()
+        {
+            return _map.size();
+        }
+
+        public Set<Map.Entry<String, Object>> getAttributes()
+        {
+            return _map.entrySet();
+        }
+
+        @Override
+        public String toString()
+        {
+            return _map.toString();
+        }
+
+        public void addAll(Attributes attributes)
+        {
+            for (String name : attributes.getAttributeNames())
+                setAttribute(name, attributes.getAttribute(name));
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return _map.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (o instanceof Attributes)
+            {
+                Attributes a = (Attributes)o;
+                for (Map.Entry<String, Object> e : _map.entrySet())
+                {
+                    if (!Objects.equals(e.getValue(), a.getAttribute(e.getKey())))
+                        return false;
+                }
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * An Attributes implementation backed by another Attributes instance, which is treated as immutable, but with a
+     * ConcurrentHashMap used as a mutable layer over it.
```

Comment:
I think using a `ConcurrentHashMap` to back the changes is useless and counter-intuitive as most (all?) method implementations aren't atomic.

I'd either change the method implementations to make them atomic (hard) or use a plain old `HashMap` as a backing  store and changing the javadoc to state that this is not thread safe, or use a lock and just state that this impl is thread-safe.

Reply:
It has to be thread safe as we are encouraging async style in the handlers, so different threads will be coordinating via request attributes.
But I agree that a concurrent losely layered over another concurrent is almost certainly wrong.    Probably need a hard atomic lock to avoid test-then-act issues.
For now, I've put in a big TODO to review.

Reply:
Since this needs to be thread-safe, I think going the `HashMap` + lock route is the best way to go as it's easy to implement. If this lock ever becomes a bottleneck, I'll be fairly simple to spend the necessary time to get rid of it.

But I'm happy with a TODO for now.